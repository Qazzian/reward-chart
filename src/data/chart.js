
const actionTypes = {
	CHART_ADD: 'CHART_ADD',
	CHART_ADD_EMOTE: 'CHART_ADD_EMOTE',
	CHART_REMOVE_EMOTE: 'CHART_REMOVE_EMOTE',
	CHART_UPDATE_EMOTE: 'CHART_UPDATE_EMOTE',
};

export function chart(chart={}, action={}) {
	switch (action.type) {
		case actionTypes.CHART_ADD:
			const {type, ...newChart} = action;
			return {
				...chart,
				...newChart,
			};
		case actionTypes.CHART_ADD_EMOTE:
			return {
				...chart,
				emotes: chart.emotes.concat(action.emoteId),
			};
		case actionTypes.CHART_REMOVE_EMOTE:
			return {
				...chart,
				emotes: chart.emotes.filter((currentEmote) => {
					return currentEmote !== action.emoteId;
				})
			};
		default:
			return chart;
	}
}

export function hasEmote(state, emoteId) {
	return state.emotes && !!(state.emotes.find((currentEmoteId) => {
		return currentEmoteId === emoteId;
	}));
}


export const chartAdd = (id, name) => ({
	type: actionTypes.CHART_ADD,
	id,
	name,
	emotes: [],
});

export const addEmote = (emoteId) => ({
	type: actionTypes.CHART_ADD_EMOTE,
	emoteId,
});

export const removeEmote = (emoteId) => ({
	type: actionTypes.CHART_REMOVE_EMOTE,
	emoteId,
});
