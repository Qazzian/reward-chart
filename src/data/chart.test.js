import * as fromChart from './chart';

const chart = fromChart.chart;

describe('Chart State', () => {

	it('should create an empty chart by default', function () {
		expect(chart()).toMatchObject({});
	});

	it('should create chart from new data', () => {
		const action = fromChart.chartAdd(1, 'test1');
		const expected = {
			name: 'test1',
			id: 1,
			emotes: [],
		};

		expect(chart({}, action)).toMatchObject(expected);
	});

	it('should add emotes', function () {
		const start = chart({}, fromChart.chartAdd(1, 'test1'));
		const action = fromChart.addEmote('test1_2018-12-14'),
			expected = {
				name: 'test1',
				id: 1,
				emotes: ['test1_2018-12-14'],
			};

		expect(chart(start, action)).toMatchObject(expected);

	});

	it('should remove existing emotes', function () {
		const startState = stateWithEmotes();
		const action = fromChart.removeEmote('test1_2018-12-15');
		expect(chart(startState, action)).toMatchSnapshot();
	});

	it('should know if it has an emote', function () {
		const startState = stateWithEmotes();
		expect(fromChart.hasEmote(startState, 'test1_2018-12-14')).toBe(true);
		expect(fromChart.hasEmote(startState, 'test1_2018-12-01')).toBe(false);
	});
});


function stateWithEmotes() {
	const emoteIds = [
		'test1_2018-12-14', 'test1_2018-12-15', 'test1_2018-12-16'
	];

	const startState = chart({}, fromChart.chartAdd(1, 'test1'));
	return emoteIds.reduce((state, emoteId) => {
		return chart(state, fromChart.addEmote(emoteId));
	}, startState);
}
