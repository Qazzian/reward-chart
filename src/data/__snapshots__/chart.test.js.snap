// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart State should remove existing emotes 1`] = `
Object {
  "emotes": Array [
    "test1_2018-12-14",
    "test1_2018-12-16",
  ],
  "id": 1,
  "name": "test1",
}
`;
